docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,122 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: ErrorState[];
+}
+
+interface ErrorState {
+    /**
+     * Locally generated unique ID for error.
+     * Allows for de-duplicating similar errors from repeated API calls.
+     */
+    id: string;
+    sdkClientMethod: 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more.
+    error: Error;
+}
+```
+
+Stateful client API method calls continue to return errors (e.g. as a failed promise). So they can be handled directly as well.
+
+The stateful clients stores the most recent errors, evicting older errors if necessary. The number of errors stored can be configured, or set to -1 to disable deletion of older errors. e.g., the composites disable error eviction in the stateful client and delete older errors themselves.
+
+The errors in the state enables UI library components that act on some errors (e.g., disabling `SendBox` in case no messages can be sent), components that surface UI errors, and adapters that generate error events for the surrounding application.
+
+### Surfacing errors to UI
+
+The ACS UI library includes an error bar component and corresponding react component bindings to surface some of the errors from the state on the UI. Surfaced errors include:
+
+* For chat:
+  * Failure to connect to ACS backend ("you are offline").
+  * Permission error when connecting to ACS backend.
+  * Permission error for a particular thread (user not in the thread).
+  * Sending messages throttled by ACS backend (temporary error).
+* For calling:
+  * Permission error when connecting to ACS backend.
```

Comment:
Calling has a huge set of codes/subcodes for why a call has ended. Are we thinking we need to account for all of these as well?

Silly question, do we need to supply "error messages" converting content from code/subcode into a string?

Reply:
This section lists the errors we'll surface in the error bar. Since this is intended for the end-user, we'll want to keep the information limited -- clearly distinguish errors that are actionable (and tell them what they can do) from those that aren't, but don't give them gory details for unactionable errors.

We will need some mapping from error code/subcode to strings to report  to the end-user
* This mapping will be part of the bindings for the error bar component.
* The total number of target strings will be limited (so we may have to pick out specific (sub)codes for some errors, but default to a generic string for most.

Doest that make sense?